clang/tools/clang-linker-wrapper/SYCLOffloadWrapper.cpp

```diff
@@ -0,0 +1,748 @@
+    return StructType::create("__tgt_offload_entry", PointerType::getUnqual(C),
+                              PointerType::getUnqual(C), getSizeTTy(),
+                              Type::getInt32Ty(C), Type::getInt32Ty(C));
+  }
+
+  // SYCL specific image descriptor type.
+  // struct __tgt_device_image {
+  //   /// version of this structure - for backward compatibility;
+  //   /// all modifications which change order/type/offsets of existing fields
+  //   /// should increment the version.
+  //   uint16_t Version;
+  //   /// the kind of offload model the image employs.
+  //   uint8_t OffloadKind;
+  //   /// format of the image data - SPIRV, LLVMIR bitcode, etc
+  //   uint8_t Format;
+  //   /// null-terminated string representation of the device's target
+  //   /// architecture
+  //   const char *DeviceTargetSpec;
+  //   /// a null-terminated string; target- and compiler-specific options
+  //   /// which are suggested to use to "compile" program at runtime
+  //   const char *CompileOptions;
+  //   /// a null-terminated string; target- and compiler-specific options
+  //   /// which are suggested to use to "link" program at runtime
+  //   const char *LinkOptions;
+  //   /// Pointer to the manifest data start
+  //   const unsigned char *ManifestStart;
+  //   /// Pointer to the manifest data end
+  //   const unsigned char *ManifestEnd;
+  //   /// Pointer to the device binary image start
+  //   void *ImageStart;
+  //   /// Pointer to the device binary image end
+  //   void *ImageEnd;
+  //   /// the entry table
+  //   __tgt_offload_entry *EntriesBegin;
+  //   __tgt_offload_entry *EntriesEnd;
+  //   _pi_device_binary_property_set_struct PropertySetBegin;
+  //   _pi_device_binary_property_set_struct PropertySetEnd;
+  // };
+  //
+  StructType *getSyclDeviceImageTy() {
+    return StructType::create(
+        {
+            Type::getInt16Ty(C),       // Version
+            Type::getInt8Ty(C),        // OffloadKind
+            Type::getInt8Ty(C),        // Format
+            PointerType::getUnqual(C), // DeviceTargetSpec
+            PointerType::getUnqual(C), // CompileOptions
+            PointerType::getUnqual(C), // LinkOptions
+            PointerType::getUnqual(C), // ManifestStart
+            PointerType::getUnqual(C), // ManifestEnd
+            PointerType::getUnqual(C), // ImageStart
+            PointerType::getUnqual(C), // ImageEnd
+            PointerType::getUnqual(C), // EntriesBegin
+            PointerType::getUnqual(C), // EntriesEnd
+            PointerType::getUnqual(C), // PropertySetBegin
+            PointerType::getUnqual(C)  // PropertySetEnd
+        },
+        "__tgt_device_image");
+  }
+
+  // SYCL specific binary descriptor type.
+  // struct __tgt_bin_desc {
+  //   /// version of this structure - for backward compatibility;
+  //   /// all modifications which change order/type/offsets of existing fields
+  //   /// should increment the version.
+  //   uint16_t Version;
```

Comment:
We have a version within a single device image and a version within this binary descriptor. Community only has a single version at the whole fat binary level - we need to align. Corresponding `TODO` comment should be added